File: 30days/day28-regex-patterns/Solution.js
// PROBLEM: https://www.hackerrank.com/challenges/30-regex-patterns/problem

'use strict';

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', inputStdin => {
    inputString += inputStdin;
});

process.stdin.on('end', _ => {
    inputString = inputString.replace(/\s*$/, '')
        .split('\n')
        .map(str => str.replace(/\s*$/, ''));

    main();
});

function readLine() {
    return inputString[currentLine++];
}



function main() {
    const N = parseInt(readLine(), 10);
    let gmailNames = [];

    for (let NItr = 0; NItr < N; NItr++) {
        const firstNameEmailID = readLine().split(' ');
        const firstName = firstNameEmailID[0];
        const emailID = firstNameEmailID[1];
        const re = new RegExp(/@gmail.com$/);

        if (re.test(emailID)) {
            gmailNames.push(firstName);
        }
    }

    gmailNames.sort();
    gmailNames.map(name => console.log(name));
}

File: 30days/day28-regex-patterns/Solution.cs
// PROBLEM: https://www.hackerrank.com/challenges/30-regex-patterns/problem

using System.CodeDom.Compiler;
using System.Collections.Generic;
using System.Collections;
using System.ComponentModel;
using System.Diagnostics.CodeAnalysis;
using System.Globalization;
using System.IO;
using System.Linq;
using System.Reflection;
using System.Runtime.Serialization;
using System.Text.RegularExpressions;
using System.Text;
using System;

class Solution {



    static void Main(string[] args) {
        int N = Convert.ToInt32(Console.ReadLine());
        List<string> gmailNames = new List<string>();

        for (int NItr = 0; NItr < N; NItr++) {
            string[] firstNameEmailID = Console.ReadLine().Split(' ');
            string firstName = firstNameEmailID[0];
            string emailID = firstNameEmailID[1];

            Regex re = new Regex(@"@gmail.com$");

            if (re.IsMatch(emailID)) {
                gmailNames.Add(firstName);
            }
        }

        gmailNames = gmailNames.OrderBy(s => s).ToList();
        
        foreach (string gmailName in gmailNames) {
            Console.WriteLine(gmailName);
        }
    }
}
